### PR TITLE
Change DB owner with ALTER AUTHORIZATION

### PR DESCRIPTION
**Pull request by robverschoor**
### Description

Implement `ALTER AUTHORIZATION` to change database ownership. 
Syntax: `ALTER AUTHORIZATION ON DATABASE::<dbname> TO <login-name>`
It is a common requirement for the DBA to create a database to be owned by a login. Thus far in Babelfish this was possible only by (temporarily)  making the target login a member of `sysadmin` role, which is undesirable.
To change the database owner to a different login, all that is required is to update `sys.babelfish_sysdatabases.owner` to the new owner's login name. There are some restrictions such as that the new owner cannot be a user in the database already (`guest` user does not count). a non-`sysadmin` DB owner cannot grant ownership to anyone else, including to a `sysadmin` login. (however, it is OK to grant ownership to yourself although there seems little point in that).
When the current DB owner has an active session in the database with the current DB set to that same database, and at the same time a different session change the database ownership, the now-previous owner retains access rights in the session as long as the database context remains unchanged; as soon as the database context is changed, the access rights from the DB ownership are lost: Babelfish behaves identically to T-SQL here.
To implement this feature, statement type `PLTSQL_STMT_CHANGE_DBOWNER` was introduced rather than something more generic-looking like `PLTSQL_STMT_ALTERAUTH`: other cases of `ALTER AUTHORIZATION` may affect the ownership/permissions-related artefacts that are created by Babelfish, or require different catalog modifications; therefore, the DB ownership change is more of a one-off case. 

Signed-off-by: Rob Verschoor [rcv@amazon.com](mailto:rcv@amazon.com)

### Issues Resolved

BABEL-2121: Babel does not support ALTER AUTHORIZATION syntax to change database owner

## Test Scenarios Covered

- Use case based - Yes

- Boundary conditions - Yes

- Arbitrary inputs - N/A

- Negative test cases - Yes

- Minor version upgrade tests - N/A

- Major version upgrade tests - N/A

- Performance tests - N/A

- Tooling impact - N/A

- Client tests - N/A

### Check List

- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: contrib/babelfishpg_tsql/src/catalog.c [hunk 1]
```diff
@@ -3250,3 +3250,76 @@ alter_guest_schema_for_db (const char *dbname)
...
+	ScanKeyData		key;
+	HeapTuple		tuple, db_found;
+	TableScanDesc	tblscan;
+		
+	Datum		values[SYSDATABASES_NUM_COLS];
+	bool		nulls[SYSDATABASES_NUM_COLS];
+	bool		replaces[SYSDATABASES_NUM_COLS];
+
+	/* Do not allow changes to system databases. */
+	/* Note: SQL Server allows changing ownership of msdb. */
```
Comment (shalinilohia50) [inline contrib/babelfishpg_tsql/src/catalog.c]: I think we should use T-SQL, not SQL Server here.

File: contrib/babelfishpg_tsql/src/pl_exec-2.c [hunk 2]
```diff
@@ -3752,3 +3752,73 @@ exec_stmt_grantschema(PLtsql_execstate *estate, PLtsql_stmt_grantschema *stmt)
...
+	}
+
+	/* Verify new owner exists as a login. */
+	if (get_role_oid(stmt->new_owner_name, true) == InvalidOid)
+	{
+		ereport(ERROR, (errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
+						errmsg("Cannot find the principal '%s', because it does not exist or you do not have permission.", stmt->new_owner_name)));
+	}
+	
+	/* SQL Server allows granting ownership to yourself when you are owner already, even without having sysadmin role. */
```
Comment (shalinilohia50) [inline contrib/babelfishpg_tsql/src/pl_exec-2.c]: I think we should use T-SQL, not SQL Server here.

File: contrib/babelfishpg_tsql/src/pl_exec-2.c [hunk 3]
```diff
@@ -3752,3 +3752,73 @@ exec_stmt_grantschema(PLtsql_execstate *estate, PLtsql_stmt_grantschema *stmt)
...
+	if (get_role_oid(stmt->new_owner_name, true) == GetSessionUserId())  // Granting ownership to myself?
+	{
+		/* Is the current login already DB owner? */
+		if (get_role_oid(get_owner_of_db(stmt->db_name), true) == GetSessionUserId())
+		{
+			granting_to_self = true;
```
Comment (shalinilohia50) [inline contrib/babelfishpg_tsql/src/pl_exec-2.c]: We can get rid of this variable  `granting_to_self` and directly add the following lines here.
```
update_db_owner(stmt->new_owner_name, stmt->db_name);
return PLTSQL_RC_OK;
```

Comment (robverschoor) [reply contrib/babelfishpg_tsql/src/pl_exec-2.c]: done

File: contrib/babelfishpg_tsql/src/pl_exec-2.c (hunk 2 above)
Comment (robverschoor) [reply contrib/babelfishpg_tsql/src/pl_exec-2.c]: done

File: contrib/babelfishpg_tsql/src/catalog.c (hunk 1 above)
Comment (robverschoor) [reply contrib/babelfishpg_tsql/src/catalog.c]: done

File: contrib/babelfishpg_tsql/src/stmt_walker.c [hunk 4]
```diff
@@ -107,6 +107,7 @@ stmt_walker(PLtsql_stmt *stmt, WalkerFunc walker, void *context)
 		case PLTSQL_STMT_INSERT_BULK:
 		case PLTSQL_STMT_SET_EXPLAIN_MODE:
 		case PLTSQL_STMT_GRANTDB:
+		case PLTSQL_STMT_CHANGE_DBOWNER:
```
Comment (HarshLunagariya) [inline contrib/babelfishpg_tsql/src/stmt_walker.c:110]: Wouldn't it be better to name for ALTERAUTH stmt rather than just change dbowner?

**Review by shalinilohia50 — APPROVED** [review]

File: contrib/babelfishpg_tsql/src/stmt_walker.c (hunk 4 above)
Comment (HarshLunagariya) [reply contrib/babelfishpg_tsql/src/stmt_walker.c:110]: Other than that functionality wise looks good.

Comment (Deepesh125) [reply contrib/babelfishpg_tsql/src/stmt_walker.c:110]: +1 for ALTERAUTH

File: contrib/babelfishpg_tsql/src/catalog.c [hunk 5]
```diff
@@ -3250,3 +3250,76 @@ alter_guest_schema_for_db (const char *dbname)
...
+	}
+
+	/* Find the database */
+	sysdatabases_rel = table_open(sysdatabases_oid, RowExclusiveLock);
+	sysdatabases_rel_descr = RelationGetDescr(sysdatabases_rel);	
+
+	ScanKeyInit(&key,
+				Anum_sysdatabases_name,
+				BTEqualStrategyNumber, F_TEXTEQ,
+				CStringGetTextDatum(db_name));
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/src/catalog.c:3288]: Do we store db_name in lowercase? if so then make sure that we pass downcases db_name here

File: test/JDBC/input/alter_authorization_change_db_owner-vu-verify.mix [hunk 6]
```diff
@@ -0,0 +1,407 @@
...
+alter server role sysadmin add member dba_login
+go
+-- end preparation
+
+-- tsql user=dba_login password=12345678
+select suser_name(), db_name()
+go
+create database change_owner_db
+go
+alter authorization on database::no_such_db_chgownerdb to new_owner_login
```
Comment (Deepesh125) [inline test/JDBC/input/alter_authorization_change_db_owner-vu-verify.mix:17]: would really like to see some case-insensitve test cases

File: contrib/babelfishpg_tsql/src/pl_exec-2.c [hunk 7]
```diff
@@ -3752,3 +3752,68 @@ exec_stmt_grantschema(PLtsql_execstate *estate, PLtsql_stmt_grantschema *stmt)
...
+	}		
+
+	/* 
+	 * The executing login must have sysadmin role: even when the current session is the owner, but has no sysadmin role, 
+	 * T-SQL does not allow the owner to grant ownership to another login -- not even to 'sa'.
+	 */
+	if (!has_privs_of_role(GetSessionUserId(), get_role_oid("sysadmin", false)))
+	{
+		ereport(ERROR, (errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
+						errmsg("Cannot find the principal '%s', because it does not exist or you do not have permission.", stmt->new_owner_name)));
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/src/pl_exec-2.c:3797]: wouldnt this error message be misleading? we should clear things out that cuttent user does not have a permission to alter authorisation on database

File: contrib/babelfishpg_tsql/src/stmt_walker.c (hunk 4 above)
Comment (robverschoor) [reply contrib/babelfishpg_tsql/src/stmt_walker.c:110]: We should keep ALTERAUTH for the more generic ALTER AUTHORIZATION case. The current case for DATABASE objects is a special subcase in that it only updates a catalog field and does not affect any of the PG role structures which have been created for the ownership/permissions structure. Those other ALTER AUTHORIZATION cases would likely be a different code path.

**Review by Deepesh125 — CHANGES_REQUESTED** [review]

File: contrib/babelfishpg_tsql/src/pl_exec-2.c (hunk 7 above)
Comment (robverschoor) [reply contrib/babelfishpg_tsql/src/pl_exec-2.c:3797]: This is the same error msg as in SQL Server

File: test/JDBC/input/alter_authorization_change_db_owner-vu-verify.mix (hunk 6 above)
Comment (robverschoor) [reply test/JDBC/input/alter_authorization_change_db_owner-vu-verify.mix:17]: see lower down in the file

File: contrib/babelfishpg_tsql/src/catalog.c (hunk 5 above)
Comment (robverschoor) [reply contrib/babelfishpg_tsql/src/catalog.c:3288]: yes, it is stored in lowercase. the db name is downcased in makeChangeDbOwnerStatement()

**Review by Deepesh125 — APPROVED** [review]

**Pull request merged**